topic_mentions: Highlight the @topic mention for topic participants.

prakhar1144 opened the pull request:
Adds support to highlight the `@topic` wildcard mention text for the users who are both:
* subscribed to the stream
* a topic participant

Topic participants: Users who either sent or reacted to the messages in the topic.

Fixes part of #22829.


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

---

**Screenshots and screen captures:**
<details><summary>Topic participant</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/7db24f46-b729-48d6-84c1-7afcddc5a052"></img>
</details> 
<details><summary>Not a topic participant</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/59d11160-1465-431b-afb7-75a06d28971d">
</img>
</details> 

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
@zulipbot add "mentor review"

abhijeetbodas2001 reviewed (commented):
Couple of comments below, but apart from that, looks good to me. Thanks!

prakhar1144 commented on web/src/stream_data.js:
```diff
@@ -864,3 +866,20 @@ export function get_options_for_dropdown_widget() {
+    for (const mssg of topic_messages) {
+        senders.push(mssg.sender_id);
+        const message_reactions = reactions.get_message_reactions(mssg);
+        reactors.push(...message_reactions.flatMap((obj) => obj.user_ids));
```
Changed it to:

```js
export function user_is_topic_participant(stream_id, topic, current_user_id) {
    // Topic participants:
    // Users who either sent or reacted to the messages in the topic.
    const topic_messages = message_util.get_messages_in_topic(stream_id, topic);

    for (const mssg of topic_messages) {
        const reactor_ids = [];
        const message_reactions = reactions.get_message_reactions(mssg);
        reactor_ids.push(...message_reactions.flatMap((obj) => obj.user_ids));

        if (current_user_id === mssg.sender_id || reactor_ids.includes(current_user_id)) {
            return true;
        }
    }

    return false;
}
```

prakhar1144 replied:
Not needed as we are using the `topic_wildcard_mentioned` flag approach now.

prakhar1144 commented on web/src/stream_data.js:
```diff
@@ -864,3 +866,20 @@ export function get_options_for_dropdown_widget() {
 }
+
+export function is_user_topic_participant(stream_id, topic, current_user_id) {
+    // Topic participants:
```
> This function sounds like it has something to do with "user topic"s, while it doesn't.

Makes sense.

**Renamed it to `user_is_topic_participant`**

> How about calling it just `is_topic_participant` (with user as the first argument to the function)

I think we shouldn't rely on arguments to understand "what a function does"?

timabbott replied:
Maybe we should have the first parameter be the user? I feel like that feels more natural to me given that name,

timabbott replied:
Or we should just have this fetch the current user ID using the function from `people.ts`, and call it `current_user_is topic_participant(stream_id, topic)`?

prakhar1144 replied:
Not needed as we are using the `topic_wildcard_mentioned` flag approach now.

timabbott commented on web/src/stream_data.js:
```diff
@@ -860,3 +862,21 @@ export function get_options_for_dropdown_widget() {
+    // Users who either sent or reacted to the messages in the topic.
+    const topic_messages = message_util.get_messages_in_topic(stream_id, topic);
+
+    for (const mssg of topic_messages) {
```
let's call the local variable here `message`.

prakhar1144 replied:
Not needed as we are using the `topic_wildcard_mentioned` flag approach now.

timabbott commented on web/src/stream_data.js:
```diff
@@ -860,3 +862,21 @@ export function get_options_for_dropdown_widget() {
         .sort((a, b) => util.strcmp(a.name.toLowerCase(), b.name.toLowerCase()));
 }
+
+export function user_is_topic_participant(stream_id, topic, current_user_id) {
```
So this function is not a correct calculation for two reasons:
* Whether the mention targets the current user for other notification purposes was determined at the time the message was sent/edited, and is somewhat recorded in the message flags; but this doesn't inspect them.
* It only looks at messages that we have locally, so older participation is not accounted for.

I wonder if there's a way to avoid needing this function? Hmm, it's possible that we need to add a `topic_mentioned` `UserMessage` flag for whether the message contains a topic mention that triggered for the current user, and then we'd be able to have a simple rule that we highlight `@topic` in a message if and only if that flag is true? Would be a data model/API change but maybe a relatively cheap one, since these should be fairly rare.

timabbott commented:
https://github.com/zulip/zulip/pull/26382#discussion_r1286203554 is probably worth a discussion in "#api design" -- after seeing this logic, my leaning is that we are probably going to want to introduce such a `UserMessage` flag such that we don't need to do an expensive and potentially inaccurate calculation when rendering individual messages (note that what you have potentially does quadratic work... since each message containing `@topic` will require us to loop through all messages in that topic!).

prakhar1144 commented:
Started a discussion: https://chat.zulip.org/#narrow/stream/378-api-design/topic/highlight.20topic-wildcard.20mention/near/1621559

prakhar1144 commented:
I'm working on this with the new approach discussed in CZO (introduce `topic_wildcard_mentioned` flag). 

I'll remove this from draft when we have something significant to review.

abhijeetbodas2001 commented on zerver/tests/test_users.py:
```diff
@@ -1861,6 +1861,7 @@ def test_stream_recipient_info(self) -> None:
             default_bot_user_ids=set(),
             service_bot_tuples=[],
             all_bot_user_ids=set(),
+            topic_participant_user_ids=set(),
         )
```
We should add a test for this fix. Basically, to check that the `wildcard_mentioned` flag is set even if notification settings are off for the user.

prakhar1144 replied:
#26506

abhijeetbodas2001 commented:
In the commit message, you have probably referenced the wrong commit. Should be 4c9d26ce17.

abhijeetbodas2001 commented:
```quote
    # IDs of topic participants who should be notified of topic wildcard mention.
    # The 'user_allows_notifications_in_StreamTopic' with 'wildcard_mentions_notify'
    # setting ON should return True.
    # A user_id can exist in either or both of the 'topic_wildcard_mention_user_ids'
    # and 'topic_wildcard_mention_in_followed_topic_user_ids' sets.
    topic_wildcard_mention_user_ids: Set[int]
```

As an aside, I think the comment for `topic_wildcard_mention_user_ids` should be similar to the one on `topic_wildcard_mention_in_followed_topic_user_ids` -- "topic participants who have followed the topic and have X setting enabled"

abhijeetbodas2001 commented:
The first commit here is a really good catch! Could you please open another PR with these 2 fixes?

prakhar1144 commented:
> The first commit here is a really good catch! Could you please open another PR with these 2 fixes?

#26506

abhijeetbodas2001 commented on zerver/tornado/event_queue.py:
```diff
@@ -618,6 +622,12 @@ def load_event_queues(port: int) -> None:
+        # `stream_wildcard_mentioned_flag` key.
+        # Remove this when one can no longer directly upgrade from 7.x to main.
+        if not hasattr(client, "stream_wildcard_mentioned_flag"):
+            client.stream_wildcard_mentioned_flag = False
```
Does this also deserve a `TODO` tag?

prakhar1144 replied:
Right. Added.

abhijeetbodas2001 reviewed (commented):
Thanks for the update!

I left some comments below.
I'm don't thoroughly understand how the events tests work with regards to client capabilities (and haven't dug into it while reviewing), so let's make sure to have another set of eyes on it.
Thanks!

prakhar1144 commented on zerver/models.py:
```diff
@@ -3403,8 +3405,9 @@ class Meta(AbstractUserMessage.Meta):
-                name="zerver_usermessage_wildcard_mentioned_message_id",
+                | Q(flags__andnz=AbstractUserMessage.flags.wildcard_mentioned.mask)
+                | Q(flags__andnz=AbstractUserMessage.flags.topic_wildcard_mentioned.mask),
+                name="zerver_usermessage_mentioned_or_wildcard_mentioned_message_id",
```
On scrolling up a bit:
```python
      models.Index(
          "user_profile",
          "message",
          condition=Q(flags__andnz=AbstractUserMessage.flags.mentioned.mask),
          name="zerver_usermessage_mentioned_message_id",
      ),
```
For personal mentions, we have the flag `mentioned`. So we have one of the indexes named `zerver_usermessage_mentioned_message_id`. To keep things consistent, I just used the term `mentioned` and not `personal_mentioned`. I think that's fine?

regarding, `stream_wildcard_mentioned or topic_wildcard_mentioned`

We collectively use the term `wildcard_mentioned` while addressing both the stream and topic wildcard mentions at other places in the codebase. 

We can definitely use `stream_wildcard_mentioned or topic_wildcard_mentioned` in the name as you already mentioned
> they don't come up anywhere in the codebase apart from migrations/models.py anyways.

@timabbott your thoughts?

abhijeetbodas2001 replied:
I definitely think it is better to more explicit here.

timabbott replied:
I think I replied in the chat.zulip.org thread; I guess reading this, I think `zerver_usermessage_mentioned_any_message_id` might be an appropriate alternative.

prakhar1144 replied:
Changed the name to `zerver_usermessage_any_mentioned_message_id`. 

ref: [CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1631746)

prakhar1144 commented on zerver/lib/message.py:
```diff
@@ -1350,10 +1363,18 @@ def apply_unread_message_event(
             user_ids_string=user_ids_string,
         )
 
+    # TODO/compatibility: Translation code for the rename of
+    # `wildcard_mentioned` to `stream_wildcard_mentioned`. Remove this
+    # when one can no longer directly upgrade from 7.x to main.
```
Added a comment, i.e.

> This is needed for the correct `stream_wildcard_mentioned in flags` check below and updating state["mentions"] accordingly.

prakhar1144 commented on zerver/lib/message.py:
```diff
@@ -254,7 +255,17 @@ def messages_for_ids(
+            flags = [
+                "wildcard_mentioned" if flag == "stream_wildcard_mentioned" else flag
+                for flag in flags
+            ]
```
Made the changes. I should have noticed this myself :)

One more point:
In `event_queue.py`, `flags` is a Collection not a list. So we can't use remove, append.

I have updated it to:
```python
      if "wildcard_mentioned" in flags:
          updated_flags = [
              "stream_wildcard_mentioned" if flag == "wildcard_mentioned" else flag
              for flag in flags
          ]
          flags = updated_flags
```

prakhar1144 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,24 @@ format used by the Zulip server that they are interacting with.
+  `client_capabilities`. When enabled, the server will send the renamed
+  flag `stream_wildcard_mentioned`, replacing `wildcard_mentioned`.
+* [`GET /events`](/api/get-events): Clients that support the
+  `stream_wildcard_mentioned_flag` client capabilities will
```
Changed this to "client capability"

prakhar1144 commented on web/tests/markdown.test.js:
```diff
@@ -733,7 +737,7 @@ test("message_flags", () => {
     assert.equal(message.is_me_message, false);
-    assert.equal(message.flags.includes("wildcard_mentioned"), true);
+    assert.equal(message.flags.includes("topic_wildcard_mentioned"), true);
     assert.equal(message.flags.includes("mentioned"), false);
```
Done

abhijeetbodas2001 replied:
Better way might be to implement a `assert_mentions(List[str])` function which asserts whatever is passed to it as is true, and whatever isn't is false.

prakhar1144 replied:
I think that would add an another layer of understanding - what `assert_mentions(List[str])` does i.e.
> it asserts whatever is passed to it as is true, and whatever isn't is false.

Currently it is easy to understand at one look.

But yeah, that is nothing super difficult to understand. Adding a comment will work.

abhijeetbodas2001 replied:
That's right. If the amount of code deduplicated seems high enough to justify implementing such a function, feel free to go ahead and implement if. Otherwise even the current situation is not too bad. I don't have a strong opinion here.

prakhar1144 commented on web/tests/markdown.test.js:
```diff
@@ -285,6 +285,10 @@ test("message_flags", () => {
+    message = {raw_content: "@**topic**"};
+    markdown.apply_markdown(message);
+    assert.ok(message.flags.includes("topic_wildcard_mentioned"));
 });
```
Makes sense. Done.

prakhar1144 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,17 @@ format used by the Zulip server that they are interacting with.
+  [`GET /messages/{message_id}`](/api/get-message): Added a new
+  flag, `topic_wildcard_mentioned`. It will be available in the
+  `flags` field of the message object for topic wildcard mentions.
+
```
Yeah, I think that would be good. Added a line:

> Now, the `wildcard_mentioned` flag will be exclusively used for stream wildcard mentions.

abhijeetbodas2001 replied:
`s/Now/Henceforth` would be better wording.

prakhar1144 commented:
@timabbott This is ready to review

Updates in the recent push:
* [After this comment](https://github.com/zulip/zulip/pull/26382#discussion_r1286203554) we pivoted to a new approach: -- [CZO Discussion : api design](https://chat.zulip.org/#narrow/stream/378-api-design/topic/highlight.20topic-wildcard.20mention/near/1621559) and [CZO Discussion : backend](https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1621615)
  * Introduce a new usermessage flag `topic_wildcard_mentioned` flag that wil be used for highlighting.
  * rename the `wildcard_mentioned` flag to `stream_wildcard_mentioned`
* I have addressed `Abhiheet Bodas`'s review comments apart from [one pending comment that needs your attention.](https://github.com/zulip/zulip/pull/26382#discussion_r1305263257)

prakhar1144 commented:
@zulipbot add "integration review"

prakhar1144 commented:
[Update 1](https://github.com/zulip/zulip/pull/26382#issuecomment-1693534097)

**Update 2:**

The recent push includes:
* Changelog feature level update -- resolved merge conflict
* Updated the index name -- [CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1631746)

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,20 @@ format used by the Zulip server that they are interacting with.
+  flag, `topic_wildcard_mentioned`. It will be available in the
+  `flags` field of the message object for topic wildcard mentions.
+  Now, the `wildcard_mentioned` flag will be exclusively used for
+  stream wildcard mentions.
```
I would probably compress these a bit more rather than duplicating so much text -- referring to the "`flags` field on message objects" ... would probably suffice to just list all 3 endpoints at the start.

prakhar1144 replied:
changelog updated as we are moving forward with a [new approach](https://chat.zulip.org/#narrow/stream/378-api-design/topic/highlight.20topic-wildcard.20mention/near/1656239)

timabbott commented on zerver/actions/message_edit.py:
```diff
@@ -248,7 +248,13 @@ def get_mentions_for_message_updates(message_id: int) -> Set[int]:
+                    | UserMessage.flags.wildcard_mentioned
+                    | UserMessage.flags.topic_wildcard_mentioned
+                )
+            )
```
Should this be a function defined in `models.py` like the `where_starred` batch of functions? If so, we could call it `where_any_mentioned`, or something.

(The benefit of putting it on `models.py` after some testing is that we'd be providing an interface that more clearly promises that we're using a query that actually uses the index we have, which is much better for future code reuse -- though we would want to test the actual query with `EXPLAIN ANALYZE` to be certain)

prakhar1144 replied:
https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1673984

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -366,7 +366,11 @@ def by_is(self, query: Select, operand: str, maybe_negate: ConditionTransform) -
+                != 0
+            )
+            cond = or_(cond1, cond2, cond3)
             return query.where(maybe_negate(cond))
```
I guess because this is SQLAlchemy, we can't use a common `models.py` function -- have you done an `EXPLAIN ANALYZE` of the SQL generated to confirm the `has:mention` narrow in fact uses this index?

timabbott replied:
OK, that query plan looks good. Though we may need to recheck this if we end up changing the index structure.

prakhar1144 replied:
https://github.com/zulip/zulip/pull/27278#discussion_r1365821523

timabbott commented on zerver/migrations/0469_add_topic_wildcard_mentioned_flag_and_create_index.py:
```diff
@@ -0,0 +1,80 @@
+                ),
+                name="zerver_usermessage_any_mentioned_message_id",
+            ),
+        ),
```
This step is the main expensive part of the migration, and it's important that this migration be applied before we restart Django into the new code, as the `is:mentioned` narrow will do a table scan without it.

I believe it can be pre-applied though, the new column is always 0 since we don't have any topic mentions yet; @alexmv FYI.

timabbott commented on web/src/compose_actions.js:
```diff
@@ -492,6 +492,7 @@ export function quote_and_reply(opts) {
 
     channel.get({
         url: "/json/messages/" + message_id,
+        data: {stream_wildcard_mentioned_flag: true},
```
Hmm, I don't love the idea that we need to remember to pass this parameter in every endpoint fetching messages.

prakhar1144 replied:
With the [updated approach,](https://chat.zulip.org/#narrow/stream/378-api-design/topic/highlight.20topic-wildcard.20mention/near/1656239) we don't need to pass these params.

timabbott commented on zerver/lib/message.py:
```diff
@@ -254,7 +255,15 @@ def messages_for_ids(
+        # this logic is intended for backwards-compatibility only.
+        if not stream_wildcard_mentioned_flag and "stream_wildcard_mentioned" in flags:
+            flags.remove("stream_wildcard_mentioned")
+            flags.append("wildcard_mentioned")
```
This may be a good question for the "api design" thread -- I wonder if it'd be better to __always__ send `stream_wildcard_mentioned`, and just have sending the `wildcard_mentioned` copy of it be the thing that's conditional on the client capability.

(The benefit of that plan would be that it'd make it easier to delete the client capability in 3 years, once there's no longer supported clients that don't support the new format).

timabbott commented on zerver/lib/message.py:
```diff
@@ -1347,10 +1358,19 @@ def apply_unread_message_event(
+    # check below and updating state["mentions"] accordingly.
+    if "wildcard_mentioned" in flags:
+        flags.remove("wildcard_mentioned")
+        flags.append("stream_wildcard_mentioned")
```
I'd be surprised if this was correct for a client that doesn't have the new client capability -- how did you verify this logic?

timabbott commented on zerver/models.py:
```diff
@@ -3294,6 +3294,7 @@ class AbstractUserMessage(models.Model):
         # Whether we've sent a push notification to the user's mobile
         # devices for this message that has not been revoked.
         "active_mobile_push_notification",
+        "topic_wildcard_mentioned",
```
Started https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1632474 to discuss this detail.

prakhar1144 replied:
```diff
-      "summarize_in_home",
-      "summarize_in_stream",
+      "topic_wildcard_mentioned",
+      "group_mentioned",
```

[Steps followed (CZO)](https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1654575)

timabbott commented:
This generally is looking pretty close to being possible to integrate; I started a few chat.zulip.org threads about a few design details that we should settle before merging this, since I think making the right decisions on them could save a lot of migration effort in the future. 

I'm pretty skeptical of the `events.py` code being correct in all cases but I also think my proposals for how to simplify how the client capability works may make it way easier to do those.

prakhar1144 commented on zerver/tornado/event_queue.py:
```diff
@@ -968,7 +968,24 @@ def is_sender_client(client: ClientDescriptor) -> bool:
+        # the `wildcard_mentioned` flag to older clients. Remove this when
+        # one can no longer directly upgrade from 7.x to main.
+        if "stream_wildcard_mentioned" in flags:
+            flags.append("wildcard_mentioned")
```
I didn't use `if "stream_wildcard_mentioned" in flags or "topic_wildcard_mentioned" in flags:` because:

If we plan to return `wildcard_mentioned = topic_wild... | stream_wild...` then, suppose during upgrade we have an event in a queue with `wildcard_mentioned` flag, now in the translation code (just above) we should replace `wildcard_mentioned` with `stream_wild..` or `topic_wild...`?

Technically it is possible to set `wildcard_mentioned` flag using @**topic**, but the feature is not yet used by web client or mobile; we also haven't mentioned it yet in our api docs. So I think it is safe to move forward with the logic that `wildcard_mentioned` should be returned only if `stream_wildcard_mentioned` is set?

For older clients  `wildcard_mentioned` is only set in the case of  stream wildcard mentions as they are unaware of "topic mentions" feature?

timabbott replied:
I think we want `if "stream_wildcard_mentioned" in flags or "topic_wildcard_mentioned" in flags`. The upgrade code path concern that you mention is generally going to affect clients for at most ~30 minutes after an upgrade -- so as long as nothing crashing, the impact of the behavior being non-ideal is very bounded, because clients will eventually reload and get a new event queue (see `server_generation` for details). If we sometimes produce an incorrect value for the flags temporarily after someone used the still-undocumented `@topic` mention code, it's not a big problem.

Whereas anyone using an old copy of the mobile apps that predates `topic_wildcard_mentioned` support, might see newly sent `@topic` messages as not containing a mention, without this code, unless I'm missing something?

timabbott replied:
I could be wrong, but I think with the present proposed API design, we may not need any compatibility code in Tornado at all -- we just need to make sure that any Django code that is generating events containing a `flags` list adds the `wildcard_mentioned` while it's adding the other flags.

prakhar1144 replied:
Yeah, passing the three fields via send_event, instead of including wildcard_mentioned in tornado code is a much cleaner approach.

prakhar1144 commented on zerver/tests/test_message_flags.py:
```diff
@@ -1072,6 +1077,11 @@ def get_unread_data() -> UnreadMessagesResult:
+        um.flags = UserMessage.flags.topic_wildcard_mentioned
+        um.save()
+        result = get_unread_data()
+        self.assertEqual(result["mentions"], [])
```
comments added

prakhar1144 commented on zerver/openapi/zulip.yaml:
```diff
@@ -896,7 +896,16 @@ paths:
+                                    This capability is for backwards-compatibility; the support for
+                                    `wildcard_mentioned` flag will be removed in a later release.
+
+                                    In Zulip 8.0 (feature level 206), added a flag
```
Merged the two commits.

prakhar1144 commented:
@timabbott 

Updates in the recent push:
* Addressed the review comments (replied to respective comments)

Important threads:
* [#backend > topic_wildcard_mentioned flag](https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1654575)
* [#api_design > Highlight topic wildcard mentions](https://chat.zulip.org/#narrow/stream/3-backend/topic/topic_wildcard_mentioned.20flag/near/1654575)

timabbott commented on zerver/migrations/0478_add_topic_wildcard_mentioned_flag_and_create_index.py:
```diff
@@ -0,0 +1,119 @@
+            clear_set_bit_topic_wildcard_and_group_mentioned,
+            reverse_code=migrations.RunPython.noop,
+            elidable=True,
+        ),
```
This sort of complex migration is generally best done in Zulip Cloud in multiple staged deployments, and for that reason we should split up this commit and its migrations. The sequence I would recommend is:
* First, the `AlterField` entries (very cheap) and `AddIndexConcurrently`
* Then, the backfill function `clear_set_bit_topic_wildcard_and_group_mentioned`.
* And finally, I think we should skip removing `zerver_usermessage_wildcard_mentioned_message_id` at all in this PR -- we will want to do it as a follow-up once we have confirmed that we're not going to have to revert this migration for some reason.

alexmv replied:
That looks like a reasonable ordering to me.  Splitting out the commit to do the backfill would also be useful to better highlight the _why_ of it -- specifically, in the first commit we can call out in the AlterField commit that we're renaming two now-unused and historical bits of the bitfield, but they may have old data that we need to clear.  The third commit can be explicit that we're clearing out old data from the now-renamed bitfields, which is why we need to walk the entire table.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6397,10 +6415,38 @@ paths:
                             like @**all**. Cannot be changed by the user directly, but
                             can change if the message is edited to add/remove
-                            a wildcard mention.
+                            a stream wildcard mention.
```
I think this documentation doesn't match my understanding of the decision, which is that this compatible `wildcard_mentioned` would be true 

Also, we always want API documentation to be written from the mindset of clearly and explicitly answering the questions that one needs to answer in order to do a correct client implementation. 

> In Zulip 8.0 (feature level 215), the `wildcard_mentioned` flag was deprecated in favor of the `stream_wildcard_mentioned` and `topic_wildcard_mentioned` flags. The `wildcard_mentioned` flag exists for backwards-compatibility with older clients, and equals `stream_wildcard_mentioned || topic_wildcard_mentioned`. Clients interacting with older servers should treat this field as an old name for `stream_wildcard_mentioned`; `@topic` mentions were not available prior to this feature level.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6397,10 +6415,38 @@ paths:
+                        <td><code>topic_wildcard_mentioned</code></td>
+                        <td>
+                            Whether this message contained topic wildcard mention i.e.
+                            @**topic**. Cannot be changed by the user directly, but
```
"Whether this message contained topic wildcard mention (`@**topic**`)" is probably clearer than the `i.e`.

prakhar1144 replied:
Done

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -245,12 +245,16 @@ def test_add_term_using_is_operator_non_dm_operand_and_negated(self) -> None:  #
 
         term = dict(operator="is", operand="mentioned", negated=True)
-        where_clause = "WHERE NOT ((flags & %(flags_1)s) != %(param_1)s OR (flags & %(flags_2)s) != %(param_2)s)"
+        where_clause = "WHERE NOT ((flags & %(flags_1)s) != %(param_1)s OR (flags & %(flags_2)s) != %(param_2)s OR (flags & %(flags_3)s) != %(param_3)s OR (flags & %(flags_4)s) != %(param_4)s)"
```
We should mainly discuss this in the `models.py` code, so feel free to mark this comment as resolved, but seeing this has been considering that we might want to write it as a single `((flags & {mention_flags_mask}))` where the mask is a bitword with the 4 relevant flags 1 and the rest 0s.

timabbott commented on zerver/migrations/0478_add_topic_wildcard_mentioned_flag_and_create_index.py:
```diff
@@ -0,0 +1,119 @@
+
+def clear_set_bit_topic_wildcard_and_group_mentioned(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
```
Let's add a blockquote here explaining that because these flags are reused flag slots in the `flags` bitword, we're not confident that their value is always in the "fresh" always-0 state in very old servers, and this migration to ensure that's the case.

prakhar1144 replied:
I have added a comment in 0486 migration file.

timabbott commented on zerver/migrations/0478_add_topic_wildcard_mentioned_flag_and_create_index.py:
```diff
@@ -0,0 +1,119 @@
+        user_profile_ids.add(user_profile_id)
+
+    user_messages = UserMessage.objects.filter(
+        user_profile_id__in=user_profile_ids, message_id__in=message_ids
```
This sort of `EXPLAIN ANALYZE` check isn't very helpful in the development environment, just because the data sizes being small can warp what index ends up being used.

I think what we want to do for this migration is something following some of the patterns in  `zerver/migrations/0239_usermessage_copy_id_to_bigint_id.py`. But roughly, we're going to want to have a loop processing batches of say 1K-10K `UserMessage` rows that match our query, ideally doing the work all in SQL, with progress indication print statements, and definitely avoid sending lists of individual message/user IDs to the database in each request, if we can.

You may want to do a pairing session with @mateuszmandera to work on this, as he has a lot of experience writing this sort of migration on our biggest tables.

alexmv replied:
Yeah.  Regardless of if that query is indexed, the fetch of the initial `ums` variable for Zulip Cloud is going to be too large to comfortably fit in memory.  In addition to being slow because it's done in Python, this also has the difficulty of exposing us to race conditions between when we fetch `user_messages` and when we update using `bulk_update`.  Any solution here is going to need to use `update` and run the logic entirely in PostgreSQL, in batches.

prakhar1144 replied:
0486 has the related migration

timabbott commented on zerver/models.py:
```diff
@@ -3562,8 +3562,10 @@ class Meta(AbstractUserMessage.Meta):
+                | Q(flags__andnz=AbstractUserMessage.flags.stream_wildcard_mentioned.mask)
+                | Q(flags__andnz=AbstractUserMessage.flags.topic_wildcard_mentioned.mask)
+                | Q(flags__andnz=AbstractUserMessage.flags.group_mentioned.mask),
+                name="zerver_usermessage_any_mentioned_message_id",
```
Let's use this as the main thread for discussing how we want this index to work... or maybe a chat.zulip.org thread pointing at this. The question in my mind is whether it would be a better index to instead index on:
`condition=Q(flags__andnz=AbstractUserMessage.flags.mentioned.mask | AbstractUserMessage.flags.stream_wildcard_mentioned.mask | AbstractUserMessage.flags.topic_wildcard_mentioned.mask | AbstractUserMessage.flags.group_mentioned.mask)`, so that we're asking the database to do a single bitmask check, rather than 4 parallel ones. 

See https://github.com/zulip/zulip/pull/26382/commits/4f550840f35ac9ee3dad9d404d2056f90f521507#r1349285843 for evidence of the precise index/query that this implementation generates.

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -968,7 +968,24 @@ def is_sender_client(client: ClientDescriptor) -> bool:
     for user_data in users:
         user_profile_id: int = user_data["id"]
-        flags: Collection[str] = user_data.get("flags", [])
+        flags: List[str] = user_data.get("flags", [])
```
Does this need to be a `Set` to avoid double-appending certain flags in the compatibility code?

prakhar1144 replied:
We no longer need these compatibility code -- [discussion](https://github.com/zulip/zulip/pull/26382#discussion_r1349311332)

timabbott commented on zerver/tornado/event_queue.py:
```diff
@@ -1075,6 +1092,17 @@ def get_client_payload(apply_markdown: bool, client_gravatar: bool) -> Dict[str,
+                "stream_wildcard_mentioned" if flag == "wildcard_mentioned" else flag
+                for flag in flags
+            ]
+            flags = updated_flags
```
Hmm, I don't think we need this sort of compatibility code. My thinking is that:
* An event that is emitted by a post-upgrade server will have all 3 fields, so this doesn't apply. (And a client that had started with a pre-8.0 server will probably just ignore the new fields, and happily access `wildcard_mentioned` until it restarts and gets an 8.0 feature level startup)
* An event that is emitted by a pre-upgrade server will just have a `wildcard_mentioned` field, and will only ever be processed by a client that had created its event queue expecting a pre-8.0 server and thus should be OK processing an additional pre-8.0 format event even if it's also getting post-8.0 events now.

prakhar1144 replied:
Yeah, passing the three fields via send_event, instead of including `wildcard_mentioned` in tornado code is a much cleaner approach.

timabbott commented:
Thanks for updating this @prakhar1144! I posted a bunch of feedback, some of which may take some work to get through. I also merged the two prep commits as the series ending with 02f3f73b0c8664624fbde2dd2bee497e048b763b.

We should do a few more revisions in this PR in order to finalize the API, Tornado code, and precise index structure, etc., but I think once that's done, we probably want to split out a prefix of this PR that is just the migrations and try to integrate that first, since that's most of the technical risk and should get extra intensive review without the clutter/complexity of a complex API change added onto it.

timabbott commented:
Let me know when this is rebased on top of https://github.com/zulip/zulip/pull/27278.

prakhar1144 commented on zerver/migrations/0486_clear_topic_wildcard_and_group_mentioned_flags.py:
```diff
@@ -0,0 +1,55 @@
+                FROM zerver_usermessage um
+                WHERE um.flags & 120 != 0
+                AND um.user_profile_id = ANY(profile_ids);
+
```
populate_db -n 10000 and ~50 personal/wildcard mentions.

```sql
zulip=> explain analyze CREATE TEMP TABLE temp_any_mentioned_usermessage AS SELECT um.id FROM zerver_usermessage um WHERE um.flags & 120 != 0 AND um.user_profile_id = ANY(ARRAY(SELECT id FROM zerver_userprofile));
                                                                        QUERY PLAN                                    
                                     
----------------------------------------------------------------------------------------------------------------------
-------------------------------------
 Bitmap Heap Scan on zerver_usermessage um  (cost=41.39..2119.44 rows=43020 width=8) (actual time=0.494..1.264 rows=28
2 loops=1)
   Recheck Cond: ((user_profile_id = ANY ($0)) AND ((flags & '120'::bigint) <> 0))
   Heap Blocks: exact=48
   InitPlan 1 (returns $0)
     ->  Seq Scan on zerver_userprofile  (cost=0.00..4.23 rows=23 width=4) (actual time=0.034..0.155 rows=23 loops=1)
   ->  Bitmap Index Scan on zerver_usermessage_any_mentioned_message_id  (cost=0.00..26.40 rows=43020 width=0) (actual
 time=0.445..0.446 rows=282 loops=1)
         Index Cond: (user_profile_id = ANY ($0))
 Planning Time: 7.418 ms
 Execution Time: 13.974 ms
(9 rows)
```

prakhar1144 commented on zerver/migrations/0486_clear_topic_wildcard_and_group_mentioned_flags.py:
```diff
@@ -0,0 +1,55 @@
+                            ORDER BY id OFFSET offset_val LIMIT batch_size
+                        )
+                    );
+
```
```sql
zulip=> explain analyze UPDATE zerver_usermessage SET flags = (flags & ~(1 << 5) & ~(1 << 6)) WHERE id = ANY(ARRAY(SELECT id FROM temp_any_mentioned_usermessage ORDER BY id OFFSET 0 LIMIT 10000));
                                                                      QUERY PLAN                                      
                                 
----------------------------------------------------------------------------------------------------------------------
---------------------------------
 Update on zerver_usermessage  (cost=164.45..241.33 rows=10 width=34) (actual time=16.909..16.913 rows=0 loops=1)
   InitPlan 1 (returns $0)
     ->  Limit  (cost=158.51..164.16 rows=2260 width=8) (actual time=0.606..0.914 rows=282 loops=1)
           ->  Sort  (cost=158.51..164.16 rows=2260 width=8) (actual time=0.603..0.722 rows=282 loops=1)
                 Sort Key: temp_any_mentioned_usermessage.id
                 Sort Method: quicksort  Memory: 38kB
                 ->  Seq Scan on temp_any_mentioned_usermessage  (cost=0.00..32.60 rows=2260 width=8) (actual time=0.0
96..0.272 rows=282 loops=1)
   ->  Index Scan using zerver_usermessage_pkey on zerver_usermessage  (cost=0.29..77.18 rows=10 width=34) (actual tim
e=1.292..3.177 rows=282 loops=1)
         Index Cond: (id = ANY ($0))
 Planning Time: 0.873 ms
 Execution Time: 17.075 ms
(11 rows)
```

prakhar1144 commented on zerver/migrations/0478_add_topic_wildcard_mentioned_flag_and_create_index.py:
```diff
@@ -0,0 +1,119 @@
+        | Q(flags__andnz=UserMessage.flags.stream_wildcard_mentioned)
+        | Q(flags__andnz=UserMessage.flags.topic_wildcard_mentioned)
+        | Q(flags__andnz=UserMessage.flags.group_mentioned)
+    ).values_list("user_profile_id", "message_id")
```
Outdated approach

prakhar1144 commented:
This is ready to review. Replied to the respective review comments.

Broadly this PR has:
* first commit: migration to clear `topic_wildcard_mentioned` and `group_mentioned` bit
* second commit: rename `wildcard_mention` to `stream_wildcard_mention` and related API changes
* third commit: frontend changes to highlight topic mentions

alya commented:
This will also close #27497 (opened recently).

alya commented:
The styling for a topic the user is not included in is incorrect. It should be greenish with no hover effect, not styled like an individual @-mention:

<img width="249" alt="Screenshot 2023-11-05 at 9 44 24 PM" src="https://github.com/zulip/zulip/assets/2090066/96f5d856-1482-4c73-bc0c-f278499be8cd">

prakhar1144 commented:
[topic-mention.webm](https://github.com/zulip/zulip/assets/56781761/4d436079-1a67-42d0-83c6-6d4cad7c4744)

>  It should be greenish with no hover effect,

@alya Do we plan to change it for both stream and topic mention?

alya commented:
Good catch! We must have failed to think about unsubscribed streams. Yes, this should apply to all wildcard mentions. This is a bug:
<img width="918" alt="Screenshot 2023-11-06 at 9 53 31 AM" src="https://github.com/zulip/zulip/assets/2090066/95f75b24-9c19-45da-b23d-dd958a9d4804">

alya commented:
That mention should be green, not bold, and with no highlighting for the message as a whole.

alya commented:
Pills should have hover effects whenever clicking on the pill opens an informational overlay (individuals, groups), and not otherwise.

prakhar1144 commented on zerver/migrations/0486_clear_topic_wildcard_and_group_mentioned_flags.py:
```diff
@@ -0,0 +1,55 @@
+    ]
+
+    operations = [
+        # Because 'topic_wildcard_mentioned' and 'group_mentioned' flags are reused flag slots (ref: 24a603b)
```
Corrected it to c37871a. Thanks.

prakhar1144 commented:
Typo fix in recent push

alexmv commented on zerver/openapi/zulip.yaml:
```diff
@@ -6551,14 +6575,43 @@ paths:
                     <tr>
                         <td><code>wildcard_mentioned</code></td>
+                        <td>
+                            Whether this message contained either
+                            <a href="/help/mention-a-user-or-group#mention-everyone-on-a-stream">stream wildcard mention</a>
+                            like @**all** or topic wildcard mention (@**topic**). Cannot be changed
```
```suggestion
                            Whether this message contained either a
                            <a href="/help/mention-a-user-or-group#mention-everyone-on-a-stream">stream wildcard mention</a>
                            like @**all** or a topic wildcard mention (@**topic**). Cannot be changed
```

alexmv commented on zerver/openapi/zulip.yaml:
```diff
@@ -6551,14 +6575,43 @@ paths:
+                    <tr>
+                        <td><code>stream_wildcard_mentioned</code></td>
                         <td>
                             Whether this message contained
```
```suggestion
                            Whether this message contained a

alexmv commented on zerver/openapi/zulip.yaml:
```diff
@@ -6551,14 +6575,43 @@ paths:
+                    <tr>
+                        <td><code>topic_wildcard_mentioned</code></td>
+                        <td>
+                            Whether this message contained topic wildcard mention (@**topic**).
```
```suggestion
                            Whether this message contained a topic wildcard mention (@**topic**).
```
Looks like we don't have a documentation we can link to yet, which I assume is a follow-up task.

prakhar1144 replied:
Yes

alexmv reviewed (commented):
This now looks good to me, modulo the tiny grammar nits.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1009,6 +1009,12 @@ class UserData(TypedDict):
+            # TODO/compatibility: The `wildcard_mentioned` flag was deprecated in favor
+            # of the `stream_wildcard_mentioned` and `topic_wildcard_mentioned` flags.
+            # The `wildcard_mentioned` flag exists for backwards-compatibility with older clients.
+            # Remove this when one can no longer directly upgrade from 7.x to main.
```
Changing the last sentence to "Remove this when we no longer support legacy clients that have not been updated to access `stream_wildcard_mentioned`."

timabbott commented on web/src/rendered_markdown.js:
```diff
@@ -137,6 +139,22 @@ export const update_elements = ($content) => {
+        // is a topic participant.
+        if (
+            stream_data.is_user_subscribed(stream_id, people.my_current_user_id()) &&
+            message.topic_wildcard_mentioned
```
I think this logic is wrong -- can't we just check `message.topic_wildcard_mentioned`? That answers the question of whether the mention applied to you or not, without any reference to subscriptions.

prakhar1144 replied:
If a user unsubscribes. Should we highlight previous topic mentions in that stream?

timabbott commented on web/src/rendered_markdown.js:
```diff
@@ -137,6 +139,22 @@ export const update_elements = ($content) => {
+        const $message_row = $content.closest(".message_row");
+        const message_id = rows.id($message_row);
+        const message = message_store.get(message_id);
+        const stream_id = message.stream_id;
```
We can't rely on `message_row` existing -- consider drafts, message edit history, and compose previews.

I'll try to rework this.

timabbott commented:
I pushed the following changes, which doesn't pass the new tests, but otherwise I think works for the logic correctness and corrects the highlighting that @alya asked for to treat this like a group mention in terms of colors (like how `@all` already works):
``` diff
diff --git a/web/src/rendered_markdown.js b/web/src/rendered_markdown.js
index c9daa9d032..e061f40a82 100644
--- a/web/src/rendered_markdown.js
+++ b/web/src/rendered_markdown.js
@@ -140,17 +140,20 @@ export const update_elements = ($content) => {
     });
 
     $content.find(".topic-mention").each(function () {
+        // TODO: This selector is designed to exclude drafts/scheduled
+        // messages. Arguably those settings should be unconditionally
+        // marked with user-mention-me, but rows.id doesn't support
+        // those elements, and we should address that quirk for
+        // mentions holistically.
         const $message_row = $content.closest(".message_row");
+        if (!$message_row.length || $message_row.closest(".overlay-message-row").length) {
+            // There's no containing message when rendering a preview.
+            return;
+        }
         const message_id = rows.id($message_row);
         const message = message_store.get(message_id);
-        const stream_id = message.stream_id;
 
-        // Highlight the mention only if the user is subscribed to the stream and
-        // is a topic participant.
-        if (
-            stream_data.is_user_subscribed(stream_id, people.my_current_user_id()) &&
-            message.topic_wildcard_mentioned
-        ) {
+        if (message.topic_wildcard_mentioned) {
             $(this).addClass("user-mention-me");
         }
     });
diff --git a/web/src/rows.js b/web/src/rows.js
index 72b54651eb..acb1d70dfa 100644
--- a/web/src/rows.js
+++ b/web/src/rows.js
@@ -70,13 +70,13 @@ export function visible_range(start_id, end_id) {
     return rows;
 }
 
-export function is_draft_row($row) {
-    return $row.find("#drafts_table .restore-overlay-message").length >= 1;
+export function is_overlay_row($row) {
+    return $row.closest(".overlay-message-row").length >= 1;
 }
 
 export function id($message_row) {
-    if (is_draft_row($message_row)) {
-        blueslip.error("Drafts have no zid");
+    if (is_overlay_row($message_row)) {
+        blueslip.error("Drafts and scheduled messages have no zid");
         return undefined;
     }
 
diff --git a/web/styles/rendered_markdown.css b/web/styles/rendered_markdown.css
index b9833964c1..f557e7f174 100644
--- a/web/styles/rendered_markdown.css
+++ b/web/styles/rendered_markdown.css
@@ -178,8 +178,7 @@
         display: inline-block;
     }
 
-    .user-mention,
-    .topic-mention {
+    .user-mention {
         color: var(--color-text-other-mention);
         background-color: var(--color-background-text-direct-mention);
 
@@ -198,12 +197,8 @@
         }
     }
 
-    .topic-mention.user-mention-me {
-        color: var(--color-text-self-group-mention);
-        background-color: var(--color-background-text-group-mention);
-    }
-
-    .user-group-mention {
+    .user-group-mention,
+    .topic-mention {
         color: var(--color-text-other-mention);
         background-color: var(--color-background-text-group-mention);
 
@@ -211,7 +206,9 @@
             color: var(--color-text-self-group-mention);
             font-weight: 600;
         }
+    }
 
+    .user-group-mention {
         &:hover {
             background-color: var(--color-background-text-hover-group-mention);
         }
diff --git a/web/tests/rendered_markdown.test.js b/web/tests/rendered_markdown.test.js
index 609ed5fbdd..48728921df 100644
--- a/web/tests/rendered_markdown.test.js
+++ b/web/tests/rendered_markdown.test.js
@@ -276,9 +276,9 @@ run_test("topic-mention", () => {
     const $content = get_content_element();
     const $mention = $.create("mention");
     $content.set_find_results(".topic-mention", $array([$mention]));
-    const $message_row = $.create(".message-row");
+    const $message_row = $.create("#message_feed_container .message-row");
     $content.closest = (closest_opts) => {
-        assert.equal(closest_opts, ".message_row");
+        assert.equal(closest_opts, "#message_feed_container .message_row");
         return $message_row;
     };
     const message_id = 100;
@@ -301,36 +301,6 @@ run_test("topic-mention", () => {
     assert.ok($mention.hasClass("user-mention-me"));
 });
 
-run_test("topic-mention not subscribed", () => {
-    // Setup
-    const $content = get_content_element();
-    const $mention = $.create("mention");
-    $content.set_find_results(".topic-mention", $array([$mention]));
-    const $message_row = $.create(".message-row");
-    $content.closest = (closest_opts) => {
-        assert.equal(closest_opts, ".message_row");
-        return $message_row;
-    };
-    const message_id = 100;
-    rows.id = (message_row) => {
-        assert.equal(message_row, $message_row);
-        return message_id;
-    };
-    const message = {
-        stream_id: 1,
-        topic_wildcard_mentioned: true,
-    };
-    message_store.get = (message_id_opt) => {
-        assert.equal(message_id_opt, message_id);
-        return message;
-    };
-    stream_data.is_user_subscribed = () => false;
-
-    assert.ok(!$mention.hasClass("user-mention-me"));
-    rm.update_elements($content);
-    assert.ok(!$mention.hasClass("user-mention-me"));
-});
-
 run_test("topic-mention not topic participant", () => {
     // Setup
     const $content = get_content_element();
@@ -354,7 +324,6 @@ run_test("topic-mention not topic participant", () => {
         assert.equal(message_id_opt, message_id);
         return message;
     };
-    stream_data.is_user_subscribed = () => false;
 
     assert.ok(!$mention.hasClass("user-mention-me"));
     rm.update_elements($content);
diff --git a/zerver/actions/message_send.py b/zerver/actions/message_send.py
index 9e343cd658..878f111a88 100644
--- a/zerver/actions/message_send.py
+++ b/zerver/actions/message_send.py
@@ -1009,10 +1009,11 @@ def do_send_messages(
         users: List[UserData] = []
         for user_id in user_list:
             flags = user_flags.get(user_id, [])
-            # TODO/compatibility: The `wildcard_mentioned` flag was deprecated in favor
-            # of the `stream_wildcard_mentioned` and `topic_wildcard_mentioned` flags.
-            # The `wildcard_mentioned` flag exists for backwards-compatibility with older clients.
-            # Remove this when one can no longer directly upgrade from 7.x to main.
+            # TODO/compatibility: The `wildcard_mentioned` flag was deprecated in favor of
+            # the `stream_wildcard_mentioned` and `topic_wildcard_mentioned` flags.  The
+            # `wildcard_mentioned` flag exists for backwards-compatibility with older
+            # clients.  Remove this when we no longer support legacy clients that have not
+            # been updated to access `stream_wildcard_mentioned`.
             if "stream_wildcard_mentioned" in flags or "topic_wildcard_mentioned" in flags:
                 flags.append("wildcard_mentioned")
             user_data: UserData = dict(id=user_id, flags=flags, mentioned_user_group_id=None)
diff --git a/zerver/lib/message.py b/zerver/lib/message.py
index abfd19abd1..e5cb1ec0e4 100644
--- a/zerver/lib/message.py
+++ b/zerver/lib/message.py
@@ -263,10 +263,11 @@ def messages_for_ids(
     for message_id in message_ids:
         msg_dict = message_dicts[message_id]
         flags = user_message_flags[message_id]
-        # TODO/compatibility: The `wildcard_mentioned` flag was deprecated in favor
-        # of the `stream_wildcard_mentioned` and `topic_wildcard_mentioned` flags.
-        # The `wildcard_mentioned` flag exists for backwards-compatibility with older clients.
-        # Remove this when one can no longer directly upgrade from 7.x to main.
+        # TODO/compatibility: The `wildcard_mentioned` flag was deprecated in favor of
+        # the `stream_wildcard_mentioned` and `topic_wildcard_mentioned` flags.  The
+        # `wildcard_mentioned` flag exists for backwards-compatibility with older
+        # clients.  Remove this when we no longer support legacy clients that have not
+        # been updated to access `stream_wildcard_mentioned`.
         if "stream_wildcard_mentioned" in flags or "topic_wildcard_mentioned" in flags:
             flags.append("wildcard_mentioned")
         msg_dict.update(flags=flags)

```

@prakhar1144 can you review those changes and fix the tests? This otherwise lgtm.

timabbott commented:
I opened https://github.com/zulip/zulip/pull/27649 to merge the database and API migrations while waiting for the frontend tests to be updated, just to manage merge conflicts/renumbering work. Huge thanks for all the work on this @prakhar1144!

alya commented:
Looks good to me!

timabbott commented:
https://github.com/zulip/zulip/pull/27649 merged successfully, so I think it's just the test failures left for this PR. I've test-deployed it on chat.zulip.org, though don't expect that to find anything until we announce the feature having merged the typeahead (https://github.com/zulip/zulip/pull/26371).

prakhar1144 commented:
Updated the tests.

Initialize Container step is failing due to network error.

timabbott commented:
Merged, thanks @prakhar1144!
